### ISSUE TITLE
Ghidra Debugger Doesn't work

### ISSUE THREAD

**Issue by dav93ide** [NationalSecurityAgency/ghidra#4275]
Hello, I'm trying to use Ghidra Debugger but it doesn't work..
I attach the debugger and start it but nothing happens, 
I just get some warning in the messages board:

<html>
<body>

jar:file:/usr/share/ghidra/Ghidra/Framework/Docking/lib/Docking.jar!/images/dialog-warning.png | Already tracking trace breakpoints
jar:file:/usr/share/ghidra/Ghidra/Framework/Docking/lib/Docking.jar!/images/dialog-warning.png | Already tracking trace breakpoints
  | Generating offer for os=GNU/Linux arch=i386:x86-64 
  | Selected first mapping offer: 
jar:file:/usr/share/ghidra/Ghidra/Framework/Docking/lib/Docking.jar!/images/dialog-warning.png | Resync: Missed loaded module/library: /usr/lib/debug/.build-id/3f/d3cdd423a18458d9db3a5e50a2b3c0d4529c07.debug | [] |
jar:file:/usr/share/ghidra/Ghidra/Framework/Docking/lib/Docking.jar!/images/dialog-warning.png | Resync: Missed loaded module/library: system-supplied DSO at 0x7ffff7fca000 | [] | Mon May 23 04:07:00 CEST 2022
jar:file:/usr/share/ghidra/Ghidra/Framework/Docking/lib/Docking.jar!/images/dialog-warning.png | Resync: Missed loaded module/library: /home/z3r0/Scrivania/All/Ghidra/ghidra_project_session_one/exercises/c1 | [] | 
jar:file:/usr/share/ghidra/Ghidra/Framework/Docking/lib/Docking.jar!/images/dialog-warning.png | Resync: Missed loaded module/library: /usr/lib/debug/.build-id/ae/60c7fd6964429befbea7c351f495224659211e.debug | [] | 
jar:file:/usr/share/ghidra/Ghidra/Framework/Docking/lib/Docking.jar!/images/dialog-warning.png | Resync: Missed unloaded module/library: /lib64/ld-linux-x86-64.so.2=agent.gdb.manager.impl.GdbModuleImpl@32678054 | [] | 
jar:file:/usr/share/ghidra/Ghidra/Framework/Docking/lib/Docking.jar!/images/dialog-warning.png | Resync: Missed unloaded module/library: /lib/x86_64-linux-gnu/libc.so.6=agent.gdb.manager.impl.GdbModuleImpl@1cb2c7cf | [] | 
  | local Pty session. PID = 189911 | [] |

On the Interpreter board I only get this message:

(gdb) 
 Temporary breakpoint 1, 0x000055555540071e in main ()

But if I click to go to the next instruction nothing happens..

What I'm doing wrong? :/
Thank you!

</body>
</html>

**Comment by d-millar** [comment]
@dav93ide Hmmm, not positive what happened here.  Any chance you clicked one of the buttons in the Threads window rather than the Objects window?  (The buttons in the Threads window control your position in the trace, whereas the buttons in the Objects window match the debugger controls.  If not, did you StepInto, StepOver, Finish, or StepLast?  Am guessing StepInto, but...  Alos, anything peculiar about your target?  Do you see the same behavior if you, say, debug /usr/bin/xclock?

**Comment by dav93ide** [comment]
Hello, No I didn't clicked on the thread menu and the program is a really simple program from a reversing exercises.
If i debug xclock and I try to attach the debugger I get this error instead:

Alert Dialog and message section with: 
Request for children timed -out - try refreshing the node.

From the interpreter: 
Function "main" not defined.
 (gdb) [Thread debugging using libthread_db enabled]
 Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Also with xclock I don't see any disassemble, but the "clock" appear!


I tryed with others really simple programs from another exercises and I get the same error as my first message. 
I can share the programs if you need them. You need to change the extension in a linux elf binary

![c2](https://user-images.githubusercontent.com/43567894/169830398-0047ffc2-74b8-4997-ae98-3f993afd4666.jpg)
![c3](https://user-images.githubusercontent.com/43567894/169830404-8183c3c4-e226-4c19-b9f2-d63a8257327f.jpg)
![c4](https://user-images.githubusercontent.com/43567894/169830406-d599d4a0-3861-4a17-bc3e-a1b0554cc5a2.jpg)

![c1](https://user-images.githubusercontent.com/43567894/169830392-0e305ac0-6da8-4134-8286-283606946504.jpg)

**Comment by dav93ide** [comment]
Sorry I tryed to upload the programs by changing the extensions but it's not possible xD

**Comment by d-millar** [comment]
@dav93ide OK, excellent - I had the right idea, but the wrong area!  So, those buttons are to help you navigate the static listing - they take you to the next Instruction, Data, Undefined, I forget what L is, Function, non-Function, etc.   You want the buttons in the Objects window that look like little yellow arrows.  (Thank you for including the screenshots - that helps Immensely!!!)

**Comment by d-millar** [comment]
P.S. I don't think any of those messages indicate a problem.  In fact, looking at your screen, I think you're in pretty good shape!

**Comment by dav93ide** [comment]
@d-millar Oh ok thank you, I was so dumb xD First time I'm trying thank you.  Now I try to step with the debugger!
Sorry if I boder you again but I get another error message with another program:
![1](https://user-images.githubusercontent.com/43567894/169836720-fd61f089-c388-453d-a017-4a5bc966c7e5.png)

In this program there's file writing, maybe is because he doensn't have the permission for the files? If so how can I give the permissions?

**Comment by dav93ide** [comment]
@d-millar I tryed to pass through using the buttons you tell me but I still get an error and if I click "resume" the program doesn't stop at the breakpoints I set. If I use single step it works thank you
![2](https://user-images.githubusercontent.com/43567894/169838078-04fa50f6-444c-46fa-be1b-eeeb248cac41.png)
![3](https://user-images.githubusercontent.com/43567894/169838083-47bafd17-118a-4d13-a1b0-fbd43bdc055e.png)
!

**Comment by d-millar** [comment]
@dav93ide No problem - everyone has to start sometime!  Regarding permissions, you may be right.  I see that "zsh 1: permesso negato" message in the Interpreter - that's a little disturbing.  If you run this program without the debugger, do you see the same message?  

On the breakpoint front, here's where things get a little complicated.  A la normal gdb, we allow you to set breakpoints in the Static listing, even if that memory isn't mapped yet.  The little pink bit in the Dynamic listing (which is above the Static listing, but almost invisible) suggests the "info proc mappings" command has failed, and we've defaulted to trying to map in all of memory (0-0x7fffffffffffffff).  That means the debugger probably doesn't know where your program lives.  If you open up the Modules panel, you can try to map the current program (in the Static listing) to the appropriate module in the Modules list (assuming you have one).  If you don't have anything in that list, let me know - we'll try to figure out how to sidestep that.

Just to confirm, maybe try running "info proc mappings" in the Interpreter window.  Once the (Static) program is associated with  (Dynamic) memory, you can either set the breakpoint in the Dynamic or the Static listing, and they should be active (red/blue? not gray).  You should also be able to click and activate any existing grayed-out breakpoints.  

If you would, if you get that error again, hit details and send us the stack track.  That'll help me narrow down what's going on there.

**Comment by d-millar** [comment]
So the easiest way to add arguments is to connect to the Agent in the Target window (which will start the agent but not the target), then Launch (not Auto-launch) from the Objects window, which allows you to enter the cmd plus parameters.  The Modules window is probably in one of the two stacks, e.g. under Breakpoints or under Regions.  The little tiny arrows allow you to navigate the stack (i.e. get to windows off screen).

**Comment by dav93ide** [comment]
Where there is the stack I can only find "Regions", "Stack" and "Console" but no Module
![1](https://user-images.githubusercontent.com/43567894/169851584-9946990a-05e1-41e7-bb23-67d654ce1391.png)
s

Do you mean the Regions window?

**Comment by d-millar** [comment]
Nope, Regions and Modules are two different things - try up to the right (?) under Breakpoints.  Also, maybe try going to the top menu and going to Windows->Debug->Modules.  Might help (although it might leave it buried) - will try it on my system in a sec.

**Comment by dav93ide** [comment]
Ok I opened the modules panel but there's nothing here :/

**Comment by d-millar** [comment]
Interesting, try "maintenance info sections" in the Interpreter and, if that works, "maintenance info sections ALLOBJ nosections".

**Comment by d-millar** [comment]
Also, maybe, is the Modules node under Inferior also empty when open?

**Comment by dav93ide** [comment]
First output command:

(gdb)maintenance info sections
Exec file:
     `/home/z3r0/Scrivania/All/Ghidra/ghidra_project_session_one/exercises/c1', file type elf64-x86-64.
  [0]      0x555555400238->0x555555400254 at 0x00000238: .interp ALLOC LOAD READONLY DATA HAS_CONTENTS
  [1]      0x555555400254->0x555555400274 at 0x00000254: .note.ABI-tag ALLOC LOAD READONLY DATA HAS_CONTENTS
  [2]      0x555555400274->0x555555400298 at 0x00000274: .note.gnu.build-id ALLOC LOAD READONLY DATA HAS_CONTENTS
  [3]      0x555555400298->0x5555554002b4 at 0x00000298: .gnu.hash ALLOC LOAD READONLY DATA HAS_CONTENTS
  [4]      0x5555554002b8->0x5555554003a8 at 0x000002b8: .dynsym ALLOC LOAD READONLY DATA HAS_CONTENTS
  [5]      0x5555554003a8->0x555555400440 at 0x000003a8: .dynstr ALLOC LOAD READONLY DATA HAS_CONTENTS
  [6]      0x555555400440->0x555555400454 at 0x00000440: .gnu.version ALLOC LOAD READONLY DATA HAS_CONTENTS
  [7]      0x555555400458->0x555555400478 at 0x00000458: .gnu.version_r ALLOC LOAD READONLY DATA HAS_CONTENTS
  [8]      0x555555400478->0x555555400538 at 0x00000478: .rela.dyn ALLOC LOAD READONLY DATA HAS_CONTENTS
  [9]      0x555555400538->0x555555400598 at 0x00000538: .rela.plt ALLOC LOAD READONLY DATA HAS_CONTENTS
  [10]     0x555555400598->0x5555554005af at 0x00000598: .init ALLOC LOAD READONLY CODE HAS_CONTENTS
  [11]     0x5555554005b0->0x555555400600 at 0x000005b0: .plt ALLOC LOAD READONLY CODE HAS_CONTENTS
  [12]     0x555555400600->0x555555400608 at 0x00000600: .plt.got ALLOC LOAD READONLY CODE HAS_CONTENTS
  [13]     0x555555400610->0x555555400842 at 0x00000610: .text ALLOC LOAD READONLY CODE HAS_CONTENTS
  [14]     0x555555400844->0x55555540084d at 0x00000844: .fini ALLOC LOAD READONLY CODE HAS_CONTENTS
  [15]     0x555555400850->0x5555554008e8 at 0x00000850: .rodata ALLOC LOAD READONLY DATA HAS_CONTENTS
  [16]     0x5555554008e8->0x555555400924 at 0x000008e8: .eh_frame_hdr ALLOC LOAD READONLY DATA HAS_CONTENTS
  [17]     0x555555400928->0x555555400a30 at 0x00000928: .eh_frame ALLOC LOAD READONLY DATA HAS_CONTENTS
  [18]     0x555555600da0->0x555555600da8 at 0x00000da0: .init_array ALLOC LOAD DATA HAS_CONTENTS
  [19]     0x555555600da8->0x555555600db0 at 0x00000da8: .fini_array ALLOC LOAD DATA HAS_CONTENTS
  [20]     0x555555600db0->0x555555600fa0 at 0x00000db0: .dynamic ALLOC LOAD DATA HAS_CONTENTS
  [21]     0x555555600fa0->0x555555601000 at 0x00000fa0: .got ALLOC LOAD DATA HAS_CONTENTS
  [22]     0x555555601000->0x555555601010 at 0x00001000: .data ALLOC LOAD DATA HAS_CONTENTS
  [23]     0x555555601010->0x555555601018 at 0x00001010: .bss ALLOC
  [24]     0x00000000->0x00000029 at 0x00001010: .comment READONLY HAS_CONTENTS



Second output command:

 (gdb)maintenance info sections ALLOBJ nosections
Exec file:
     `/home/z3r0/Scrivania/All/Ghidra/ghidra_project_session_one/exercises/c1', file type elf64-x86-64.

**Comment by dav93ide** [comment]
After I ran the command the module window populated:

![1](https://user-images.githubusercontent.com/43567894/169860783-c63bbff5-c6ea-4438-a273-cdbde41dc70b.png)

**Comment by d-millar** [comment]
More to the point, you should be able to enable a breakpoint, and clicking on any of those base addresses should bring up the appropriate bytes in the Dynamic Listing.  Looks like c1 is not in the modules list - you might have to step once and refresh the Modules list from the tree.  (Checking, I think this is a know issue for some gdb targets.)

**Comment by dav93ide** [comment]
@d-millar Hello sorry for late answer I was busy.
I retryed the debugger but when I setup a breakpoint and I click the >> it still doesn't stop at the breakpoint, he just pass it and go to the end of the program.
But if I single step the instructions it works right. What should I do?

**Comment by dav93ide** [comment]
Also only the "Step Into" button functions, all others like "Step Over", "Finish" and "Step Last" don't work.

**Comment by d-millar** [comment]
@dav93ide No worries as far as the "late answer".  We're here and any answer helps us out - hopefully you too.

OK, so on the problem at hand, in your second screenshot, as you I think realized, the program has run to completion and so everything is disabled.  You can see this from the fact that the process in the Objects window is greyed out, as well as the "inferior exited" messages.  So, obviously, missed the breakpoint.

I believe this is because Ghidra is still not making the association between the program in the static listing and the running target process.  There are several possible solutions for this.  The easiest, albeit least satisfying, solution is to set the breakpoint in the Interpreter at the dynamic address, i.e. "break 0x55555540071e".  I think that will guarantee it's enabled.  Seting the breakpoint in the Dynamic listing "should" do the same, i.e. if you go to 0x55555540071e (G -> 0x55555540071e:8) and set the breakpoint (K), the breakpoint should be active.  You can verify that it's been set using "info breakpoints" in the Interpreter or making sure the icons in the Breakpoint window are not grey or split.  I would definitely be interested if this does not work.

The third thing I would try is to do exactly what you've already done, but make sure the breakpoint is enabled.  You can do this, as mentioned, using raw gdb commands in the Interpreter, by clicking the greyed-out breakpoints in the Breakpoint window, or by clicking that tiny little button in the Messages line that says, "There are ineffective breakpoints that can be placed."  Basically, it looks like you did everything correctly, but, for reasons that are still unclear to me, the breakpoint was not activated.  That is to say, it was associated with the program in gdb and Ghidra, but not with the running process.

Regarding the other stepping commands, in "main" you have to be a little careful.  "Finish" is the equivalent of "end the program" from main, as there's nowhere to go from the current stack location.  "Step over' should work, but, if you step over a call that terminated the program without returning to the instruction after the call, whether because of the normal control flow or an exception handler, you're out of luck again.  ("Step last" just repeats whatever step command you did last.)

**Comment by dav93ide** [comment]
Hello, thank you for your efforts and your time!
Now it looks like it works!
But I think I didn't understand the "Third Thing", do you mean to right-click on the breakpoint in the breakpoints window and then click on "enable breakpoint"?

Sorry about that but I'm a little noob on disassembly ecc, I'm still learning (:
Anyway I'm really enjoying using Ghidra, you're doing a really good job on this tool!

**Comment by d-millar** [comment]
@dav93ide  Yep, exactly.  If the breakpoint is only present in the Static listing (and, correspondingly, the top half of the Breakpoints window), you can enable in at least four different ways:  (1) as you mentioned, right-click and select "Enable", (2) click (or possibly double-click) the icon in the top half of the Breakpoints window, (3) click the icon in the Static listing, or (4) click the button embedded in the "There are ineffective breakpoints..." message in the Debug Console.  There are technically two cases you could be in here with the breakpoint disabled - in the first, the breakpoint will be present in the top half of the Breakpoints window and the Static listing (which typically is true before you've started the debugger or if you had breakpoints in the program before you started the debugger), or, in the second, the breakpoint will be mirrored in the Static and Dynamic listings and the top and bottom halves of the Breakpoint viewer but not a solid colored circle (typically the case if you or the debugger has decided to disable the breakpoint).  A little confusing, I'll admit - am blaming gdb.  :)

Glad you're enjoying the tool and learning about RE (me too) - always interested in any suggestions you might have for improving it!

**Comment by d-millar** [comment]
@dav93ide  I'm going to move this issue to Discussions because I think it may help other folks down the line.  Feel free to continue the conversation here (or file a new ticket) if you have new questions.
